test/integration/bindings/fixture-navigation/toolpad/pages/page1/page.yml

```diff
@@ -12,7 +12,8 @@ spec:
           props:
             onClick:
               $$navigationAction:
-                page: au03sef
+                page:
+                  $ref: au03sef
```

Comment:
This doesn't seem right. `page` is supposed to be a string according to the schema: https://github.com/mui/mui-toolpad/blob/ae82d58325f769644a7fb0f2f0b6b4683e75d285/packages/toolpad-app/src/server/schema.ts#L58
how is this passing the validation? I'd expect a test to break because of this

Reply:
I think the code itself was expecting a ref, anyway using a string seems good - I'll make it a string and take a look at the schema validation.

Reply:
It's not easy to make the schema validation work here, seems like `z.record(jsonSchema)` is allowing any object to pass validation, but without it the unions also don't work as expected. So far haven't solved it, I'm looking into it.

Reply:
ok i finally found a way

Reply:
Aah good catch,

One thing we could do is disallow properties in the `jsonSchema` that start with `$$`

```tsx
const literalSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
type Literal = z.infer<typeof literalSchema>;
type Json = Literal | { [key: string]: Json } | Json[];
export const jsonSchema: z.ZodType<Json> = z
  .lazy(() =>
    z.union([
      ...literalSchema.options,
      z.array(jsonSchema),
      // disallow $$ properties here
      z.record(jsonSchema).refine((obj) => !Object.keys(obj).some((key) => key.startsWith('$$')))
    ]),
  )
```

But this would ofcourse not work for nested bindings, so we need to pull back in `jsExpressionBindingSchema`. But then we should probably not call it `jsonSchema` anymore. We can probably instead remove `jsonSchema` from `bindablePropSchema` and insteadcreate that type like. (You'll have to do some trickery to make recursive types work, see how it's done in `jsonSchema`)

```tsx
export const bindablePropSchema = z.lazy(() => z.union([
  ...literalSchema.options,
  z.array(bindablePropSchema),
  z.record(bindablePropSchema).refine((obj) => !Object.keys(obj).some((key) => key.startsWith('$$'))),
  jsExpressionBindingSchema,
  envBindingSchema,
  jsExpressionActionSchema,
  navigationActionSchema,
  templateSchema,
]));

```

Reply:
Yep this exactly what I was already going, I was just going to try to pass all the possible `$$...` properties to exclude in the refine function, but we can also just exclude if they start with `$$`, maybe that's easier to maintain.